_posts/2021-06-16-android-flexible-configuration-with-deferred-resources.md

```diff
@@ -0,0 +1,238 @@
+---
+title: "Android: Flexible configuration with Deferred Resources"
+excerpt: Deferred Resources provides flexibility beyond the standard resource-resolution approaches.
+tags: Mobile Android Open-source Deferred-Resources
+authors:
+- Drew Hamilton
+header:
+  teaser: /assets/images/post/05_color_configuration.png
+  teaser_alt: Deferred Resources
+category: Mobile
+---
+
+Our feature libraries, which include UI, have a number of varying requirements regarding
+configuration by our customers. (My colleague Hari wrote about one such library's requirements
+[here]({% post_url 2020-08-14-android-configuration-driven-ui-from-epoxy-to-compose %}).) One
+category of common configurations is resources—text, colors, images—which may be defined in various
+ways, like code, resources, or theme attributes. We created
+[Deferred Resources](https://engineering.backbase.com/DeferredResources) to support such
+configurations, and with it our customers can consistently declare configuration properties from
+any of these common sources:
+```kotlin
+SomeConfiguration {
+    textColor = DeferredColor.Constant(Color.WHITE)
+    // or
+    textColor = DeferredColor.Resource(R.color.splash_text)
+    // or
+    textColor = DeferredColor.Attribute(R.attr.colorOnSurface)
+}
+
+// Resolved to a @ColorInt in our UI:
+@ColorInt val textColor = configuration.textColor.resolve(context)
+```
+
+This covers most of our customers' use cases. Deferred Resources has been working well for us for
+over a year now.
+
+But beyond these common use cases, Deferred Resources' design provides a lot of flexibility to
+provide resources in other ways. Each deferred resource type is an interface with one or more
+abstract functions to resolve the underlying resource. Thus, a user of the Deferred Resources
+library can define any resource-resolution implementation they'd like. The following are a few
+examples of how we are taking advantage of this flexibility.
+
+## Color variants
+
+The Backbase design system has a concept of "color variants," where any theme color has lighter and
+darker alternates. These variants are defined by a computed overlay: a "lighter" variant is the
+base color with a 30% white overlay, while a "darker" variant is the base color with a 30% black
+overlay.
+
+```kotlin
+enum class ColorVariant(
+    @ColorInt internal val overlay: Int
+) {
+    LIGHTER(overlay = Color.WHITE.withAlpha(0x4D)),
+    DARKER(overlay = Color.BLACK.withAlpha(0x4D)),
+}
+```
+
+By shipping a `DeferredVariantColor` class with our design system, we can make it very easy for our
+feature libraries as well as for our customers to use the same variants:
+
+```kotlin
+/**
+ * Convert a [DeferredColor] to a [variant] of the same color without resolving it
+ * yet.
+ */
+public fun DeferredColor.variant(variant: ColorVariant): DeferredColor =
+  DeferredVariantColor(this, variant)
+
+@Poko internal class DeferredVariantColor(
```

Comment:
Maybe indicate somewhere in this article what Poko is

Reply:
Yeah I was debating this. Worried it would distract from the content. Maybe a footnote at the bottom?

Reply:
Or how about a code comment above this line:
```kotlin
// See https://github.com/drewhamilton/Poko for more on the @Poko annotation
```

Reply:
Added the comment, let me know what you think.

Reply:
Yeah a code comment is fine I think